crates/wasmtime/src/compiler.rs

```diff
@@ -0,0 +1,700 @@
+            };
+
+            if x.key.kind() == CompileKey::WASM_FUNCTION_KIND
+                || x.key.kind() == CompileKey::ARRAY_TO_WASM_TRAMPOLINE_KIND
+                || x.key.kind() == CompileKey::NATIVE_TO_WASM_TRAMPOLINE_KIND
+            {
+                indices
+                    .compiled_func_index_to_module
+                    .insert(index.unwrap_function(), x.key.module());
+                if let Some(info) = x.info {
+                    indices.wasm_function_infos.insert(x.key, info);
+                }
+            }
+
+            indices
+                .indices
+                .entry(x.key.kind())
+                .or_default()
+                .insert(x.key, index);
+        }
+        (compiled_funcs, indices)
+    }
+}
+
+#[derive(Default)]
+pub struct FunctionIndices {
+    // A reverse map from an index in `compiled_funcs` to the
+    // `StaticModuleIndex` for that function.
+    compiled_func_index_to_module: HashMap<usize, StaticModuleIndex>,
+
+    // A map from Wasm functions' compile keys to their infos.
+    wasm_function_infos: HashMap<CompileKey, WasmFunctionInfo>,
+
+    // The index of each compiled function, bucketed by compile key kind.
+    indices: BTreeMap<u32, BTreeMap<CompileKey, CompiledFunction<usize>>>,
+}
+
+impl FunctionIndices {
+    /// Link the compiled functions together, resolving relocations, and append
+    /// them to the given ELF file.
+    pub fn link_and_append_code<'a>(
+        mut self,
+        mut obj: object::write::Object<'static>,
+        tunables: &'a Tunables,
+        compiler: &dyn Compiler,
+        compiled_funcs: Vec<(String, Box<dyn Any + Send>)>,
+        translations: PrimaryMap<StaticModuleIndex, ModuleTranslation<'_>>,
+    ) -> Result<(wasmtime_jit::ObjectBuilder<'a>, Artifacts)> {
+        // Append all the functions to the ELF file.
+        //
+        // The result is a vector parallel to `compiled_funcs` where
+        // `symbol_ids_and_locs[i]` is the symbol ID and function location of
+        // `compiled_funcs[i]`.
+        let symbol_ids_and_locs = compiler.append_code(
+            &mut obj,
+            &compiled_funcs,
+            tunables,
+            &|caller_index: usize, callee_index: FuncIndex| {
+                let module = self
+                    .compiled_func_index_to_module
+                    .get(&caller_index)
+                    .copied()
+                    .expect("should only reloc inside wasm function callers");
+                let def_func_index = translations[module]
+                    .module
+                    .defined_func_index(callee_index)
+                    .unwrap();
+                self.indices[&CompileKey::WASM_FUNCTION_KIND]
+                    [&CompileKey::wasm_function(module, def_func_index)]
+                    .unwrap_function()
+            },
+        )?;
+
+        // If requested, generate and add dwarf information.
+        if tunables.generate_native_debuginfo {
```

Comment:
This'll want a gate that there's only one translation because otherwise appending dwarf multiple times will end up with something corrupt that nothing can read